src/par_iter/find_first_last.rs

```diff
@@ -0,0 +1,183 @@
+use std::cell::Cell;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use super::internal::*;
+use super::*;
+use super::len::*;
+
+// The consumer for find_first/find_last has fake indexes representing the lower
+// and upper bounds of the "range" of data it consumes. This range does not
+// correspond to indexes from the consumed iterator but rather indicate the
+// consumer's position relative to other consumers. The purpose is to allow a
+// consumer to know it should stop consuming items when another consumer finds a
+// better match.
+
+// An indexed consumer could specialize to use the real indexes instead, but we
+// don't implement that for now. The only downside of the current approach is
+// that in some cases, iterators very close to each other will have the same
+// range and therefore not be able to stop processing if one of them finds a
+// better match than the others.
```

Comment:
Thanks, I'll work on that explanation.

I think `AtomicU64` is not stable yet (tracked [here](https://github.com/rust-lang/rust/issues/32976)). Would it be worth postponing merging this PR until it's stable? Of course, we should be able to make that change later, because the exact size of the integer used is private to this module.

I agree that a test would be useful to validate my hand-waving in the comments. Any thoughts on how to write that test? I don't yet know enough about the splitting logic to know how to force that case. One possibility is to start with a `FindConsumer` whose upper and lower bounds are the same, but that requires either making the `FindConsumer` struct public so it can be used in the test module, or writing a test directly in this module (which none of the rest of this project does).